discovery_engine/lib/src/domain/event_handler.dart

```diff
@@ -0,0 +1,216 @@
+    show activeDocumentDataBox, changedDocumentIdBox, documentBox;
+import 'package:xayn_discovery_engine/src/infrastructure/repository/hive_active_document_repo.dart'
+    show HiveActiveDocumentDataRepository;
+import 'package:xayn_discovery_engine/src/infrastructure/repository/hive_changed_document_repo.dart'
+    show HiveChangedDocumentRepository;
+import 'package:xayn_discovery_engine/src/infrastructure/repository/hive_document_repo.dart'
+    show HiveDocumentRepository;
+import 'package:xayn_discovery_engine/src/infrastructure/type_adapters/hive_duration_adapter.dart'
+    show DurationAdapter;
+import 'package:xayn_discovery_engine/src/infrastructure/type_adapters/hive_unique_id_adapter.dart'
+    show DocumentIdAdapter, StackIdAdapter;
+import 'package:xayn_discovery_engine/src/infrastructure/type_adapters/hive_uri_adapter.dart'
+    show UriAdapter;
+import 'package:xayn_discovery_engine/src/logger.dart' show logger;
+
+const kEnginePath = 'engine';
+const kDatabasePath = 'database';
+
+class EventHandler {
+  Future<Engine>? _engineFuture;
+  late final DocumentRepository _documentRepository;
+  late final ActiveDocumentDataRepository _activeDataRepository;
+  late final ChangedDocumentRepository _changedDocumentRepository;
+  late final DocumentManager _documentManager;
+  late final FeedManager _feedManager;
+
+  /// Decides what to do with incoming [ClientEvent] by passing it
+  /// to a dedicated manager and returns the appropriate response in the form
+  /// of a proper [EngineEvent].
+  ///
+  /// This handler is invoked by a [ClientEvent]s stream listener, so it is
+  /// called each time there is a new event on the stream, without waiting for
+  /// previous events to finish processing.
+  Future<EngineEvent> handleMessage(ClientEvent clientEvent) async {
+    if (clientEvent is Init) {
+      try {
+        _engineFuture = _initEngine(clientEvent.configuration);
+        await _engineFuture;
+        return const EngineEvent.clientEventSucceeded();
+      } catch (e) {
+        logger.e(e);
+        _engineFuture = null;
+        final reason = e is AssetFetcherException
+            ? EngineExceptionReason.failedToGetAssets
+            : EngineExceptionReason.genericError;
+        return EngineExceptionRaised(reason);
+      }
+    }
+
+    if (_engineFuture == null) {
+      return const EngineEvent.engineExceptionRaised(
+        EngineExceptionReason.engineNotReady,
+      );
+    }
+
+    try {
+      await _engineFuture;
+    } catch (_) {
+      return const EngineEvent.engineExceptionRaised(
+        EngineExceptionReason.engineNotReady,
+      );
+    }
+
+    // prepare reposnses
+    EngineEvent response = const EngineEvent.clientEventSucceeded();
+
+    try {
+      if (clientEvent is FeedClientEvent) {
+        response = await _feedManager.handleFeedClientEvent(clientEvent);
+      } else if (clientEvent is DocumentClientEvent) {
+        await _documentManager.handleDocumentClientEvent(clientEvent);
+      }
```

Comment:
should we handle the `else` case as well, e.g. by returning an error?

Reply:
I added additional branch for system events and an else branch too, but I think handling system events like `ResetEngine` and `ConfigurationChanged` is beyond the scope of this PR.

Reply:
I agree, leaving a TODO in there is fine for now.